[a2av] Add make_exchange_plan for reuse of meta data

kwen2501 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #168047
* #167792
* #167676
* #167144
* #166896
* #166899
* #166898
* #166774
* #163815
* #163814
* #166772
* #165482
* #165101
* #164181
* __->__ #164164

Adding a new API to exchange all-to-all-v splits and offsets.

```
in_splits = ...
plan = dist.symm_mem.make_a2a_exchange_plan(
    in_splits, src_offsets, out_splits, dst_offsets, group_name
)
```

The end goal is so that users can reuse the `plan` to exchange multiple data following the same splits:
```
all_to_all_v(topk_idx_in, topk_idx_out, plan)
all_to_all_v(topk_weights_in, topk_weights_out, plan)
all_to_all_v(tokens_in, tokens_out, plan)
```

cc @H-Huang @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @pragupta @msaroufim @dcci

kwen2501 committed:
Update

[ghstack-poisoned]

kwen2501 committed:
Update

[ghstack-poisoned]

kwen2501 committed:
Update

[ghstack-poisoned]

kwen2501 commented on torch/distributed/_symmetric_memory/__init__.py:
```diff
@@ -1824,4 +1824,30 @@ def get_mempool_allocator(device: _device):  # type: ignore[no-untyped-def]
+    out_splits: torch.Tensor,
+    dst_offsets: torch.Tensor,
+    group_name: str,
+) -> tuple[torch.Tensor, torch.Tensor, torch.Tensor, torch.Tensor]:
```
Thanks good idea. Added.

kwen2501 committed:
Update

[ghstack-poisoned]

kwen2501 commented on torch/csrc/distributed/c10d/symm_mem/nvshmem_extension.cu:
```diff
@@ -332,13 +338,16 @@ void all_to_all_vdev(
+  // Use collective launch because kernel involves nvshmem barrier
   nvshmemx_collective_launch(
-      (const void*)exchangeSplitAndOffset,
+      (const void*)exchangeSplitAndOffsetKernel,
```
Will refactor all instances in follow-up PR. Thanks!

kwen2501 committed:
Update

[ghstack-poisoned]

Skylion007 commented on torch/distributed/_symmetric_memory/__init__.py:
```diff
@@ -1824,4 +1825,49 @@ def get_mempool_allocator(device: _device):  # type: ignore[no-untyped-def]
+- out_splits: splits of peers' contribution to my output.
+- dst_offsets: offsets within my output where I should store peers' contribution.
+"""
+ExchangePlan = namedtuple(
```
Please use Typed NamedTuple API. This version is deprecated in Python anyway.

ngimel reviewed (approved):
Looks good, could you switch to NamedTuple as @Skylion007 suggests?

kwen2501 commented on test/distributed/test_nvshmem.py:
```diff
@@ -557,6 +557,66 @@ def test_all_to_all_vdev_2d_offset(self) -> None:
+        src_offsets = symm_mem.empty(
+            self.world_size, dtype=torch.int64, device=self.device
+        )
+        out_splits = symm_mem.empty(
```
Yes :)

kwen2501 committed:
Update on "[a2av] Add make_exchange_plan for reuse of meta data"


Adding a new API to exchange all-to-all-v splits and offsets.

```
in_splits = ...
plan = dist.symm_mem.make_a2a_exchange_plan(
    in_splits, src_offsets, out_splits, dst_offsets, group_name
)
```

The end goal is so that users can reuse the `plan` to exchange multiple data following the same splits:
```
all_to_all_v(topk_idx_in, topk_idx_out, plan)
all_to_all_v(topk_weights_in, topk_weights_out, plan)
all_to_all_v(tokens_in, tokens_out, plan)
```

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta msaroufim dcci

[ghstack-poisoned]
